core-processor/src/ext.rs

```diff
@@ -1199,15 +1199,37 @@ impl<LP: LazyPagesInterface> Externalities for Ext<LP> {
     }
 
     fn unreserve_gas(&mut self, id: ReservationId) -> Result<u64, Self::FallibleError> {
-        let amount = self.context.gas_reserver.unreserve(id)?;
+        let (amount, reimburse) = self.context.gas_reserver.unreserve(id)?;
 
-        // This statement is like an op that increases "left" counter, but do not affect "burned" counter,
-        // because we don't actually refund, we just rise "left" counter during unreserve
-        // and it won't affect gas allowance counter because we don't make any actual calculations
-        // TODO: uncomment when unreserving in current message features is discussed
-        /*if !self.context.gas_counter.increase(amount) {
-            return Err(some_charge_error.into());
-        }*/
+        if let Some(reimbursement) = reimburse {
+            let current_gas_amount = self.gas_amount();
+
+            // Basically amount of the reseravtion and the cost for the hold duration.
+            let reimbursement_amount = self
+                .context
+                .costs
+                .rent
+                .reservation
+                .cost_for(
+                    self.context
+                        .reserve_for
+                        .saturating_add(reimbursement.duration())
+                        .into(),
+                )
+                .saturating_add(amount);
+            self.context
+                .gas_counter
+                .increase(reimbursement_amount, reimbursement)
+                .then_some(())
+                .unwrap_or_else(|| {
+                    unreachable!(
```

Comment:
I guess `log::error!()` should be here too?

Reply:
done